backend/contracts/OasisReward.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+import "@openzeppelin/contracts/utils/Base64.sol";
+
+contract OasisReward is ERC721Enumerable {
+    // Mapping from token ID to IPFS URI
+    mapping(uint256 => string) private _tokenURIs;
+
+    // Mapping from user address to list of owned token IDs
+    mapping(address => uint256[]) private _ownedTokens;
+
+    // Contract owner
+    address private _owner;
+
+    
+    modifier onlyOwner {
+        require(tx.origin == _owner, "Owner address not tx.origin");
+        _;
+    }
+
+    constructor(
+        string memory name,
+        string memory symbol
+    ) ERC721(name, symbol) {
+        _owner = msg.sender;
+    }
+
+    /**
+     * @notice Mint new NFT reward
+     * @param to address to mint the NFT to
+     * @param base64EncodedSVG base64 encoded SVG image
+     */
+    // Function to mint new tokens
+    function mint(address to, string calldata base64EncodedSVG) public onlyOwner {
```

Comment:
Is `onlyOwner` the correct modifier here? I think the caller (quiz) should be checked, if it's whitelisted, because you can have multiple quizzes using the same NFT contract. Sth like:

```suggestion
    mapping(address => bool) _allowMint;
    
    modifier onlyAllowMint {
      require(_allowMint[msg.sender], "not allowed to mint");
      _;
    }
    
    function addAllowMint(address) onlyOwner {
      ...
    }
    
    function removeAllowMint(address) onlyOwner {
      ...
    }
    
    function mint(address to, string calldata base64EncodedSVG) public onlyAllowMint {
```

Reply:
You're right, the tx.origin check does not take into account user mints and gasless tx, will change to whitelisted addresses.